common/djangoapps/student/management/commands/bulk_unenroll.py

```diff
@@ -57,36 +57,42 @@ def unenroll_users_from_csv(self, csv_file):
         Un-enroll the given set of users provided in csv
         """
         reader = list(unicodecsv.DictReader(csv_file))
-        users_unenrolled = {}
-        for row in reader:
-            username = row['username']
-            course_key = row['course_id']
-
-            try:
-                course_id = CourseKey.from_string(row['course_id'])
-            except InvalidKeyError:
-                msg = 'Invalid course id {course_id}, skipping un-enrollement for {username}, {email}'.format(**row)
-                logger.warning(msg)
-                continue
-
-            try:
-                enrollment = CourseEnrollment.objects.get(user__username=username, course_id=course_id)
-                enrollment.update_enrollment(is_active=False, skip_refund=True)
-                if username in users_unenrolled:
-                    users_unenrolled[username].append(course_key.encode())
-                else:
-                    users_unenrolled[username] = [course_key.encode()]
-
-            except ObjectDoesNotExist:
-                msg = 'Enrollment for the user {} in course {} does not exist!'.format(username, course_key)
-                logger.info(msg)
-
-            except Exception as err:  # pylint: disable=W0703
-                msg = 'Error un-enrolling User {} from course {}: with error: {}'.format(username, course_key, err)
-                logger.error(msg, exc_info=True)
-
-        logger.info("Following users have been unenrolled successfully from the following courses: {users_unenrolled}"
-                    .format(users_unenrolled=["{}:{}".format(k, v) for k, v in users_unenrolled.items()]))
+        column_names = [name for name in reader[0].keys()]
+        if len(column_names) == 1 and 'course_id' in column_names:
+            for row in reader:
+                self.unenroll_all_users(course_id=row['course_id'])
+            return
+        else:
+            users_unenrolled = {}
+            for row in reader:
+                username = row['username']
+                course_key = row['course_id']
+
+                try:
+                    course_id = CourseKey.from_string(row['course_id'])
+                except InvalidKeyError:
+                    msg = 'Invalid course id {course_id}, skipping un-enrollement for {username}, {email}'.format(**row)
+                    logger.warning(msg)
+                    continue
+
+                try:
+                    enrollment = CourseEnrollment.objects.get(user__username=username, course_id=course_id)
+                    enrollment.update_enrollment(is_active=False, skip_refund=True)
```

Comment:
can we use `update_enrollment()` in `unenroll_all_users()` as well so that it can emit the unenroll events and update the request cache?

Reply:
Also move this code into a helper function and use it here and in `unenroll_all_users()`.

Reply:
I have suggested like following so that we can avoid duplication of code:
```
def unenroll_learners(self, course_id, username=None):
        """
        Un-enroll learners
        """
        try:
            course_id = CourseKey.from_string(course_id)
        except InvalidKeyError:
            msg = 'Invalid course id {}, skipping un-enrollement.'.format(course_id)
            logger.warning(msg)
            return

        filter_query = (Q(course_id=course_id) & Q(is_active=True))
        if username:
            filter_query.add(Q(user__username=username))

        enrollment_qs = CourseEnrollment.objects.filter(filter_query)
        for enrollment in enrollment_qs:
            enrollment.update_enrollment(is_active=False, skip_refund=True)
            logger.info("User [{}] have been successfully unenrolled from the course: {}"
                    .format(enrollment.user.username, course_id))
```

Reply:
And use this function as needed.

Reply:
Thanks for sharing the snippets!